protocol/x/clob/keeper/process_operations.go

```diff
@@ -248,6 +214,218 @@ func (k Keeper) PersistMatchToState(
 	return nil
 }
 
+// PersistOrderRemovalToState takes in an OrderRemoval, statefully validates it according to
+// RemovalReason, and writes the removal to state.
+func (k Keeper) PersistOrderRemovalToState(
+	ctx sdk.Context,
+	orderRemoval *types.OrderRemoval,
+) error {
+	orderIdToRemove := orderRemoval.GetOrderId()
+	orderIdToRemove.MustBeStatefulOrder()
+
+	// Order removals are always for stateful orders that must exist.
+	orderToRemove, err := k.FetchOrderFromOrderId(ctx, orderIdToRemove, make(map[types.OrderId]types.Order, 0))
+	if err != nil {
+		return err
+	}
+
+	// Statefully validate that the removal reason is valid.
+	switch removalReason := orderRemoval.RemovalReason; removalReason {
+	case types.OrderRemoval_REMOVAL_REASON_UNDERCOLLATERALIZED:
+		// For the collateralization check, use the remaining amount of the order that is resting on the book.
+		remainingAmount, hasRemainingAmount := k.MemClob.GetOrderRemainingAmount(ctx, orderToRemove)
+		if !hasRemainingAmount {
+			panic(fmt.Sprintf("addOrderToOrderbookCollateralizationCheck: order has no remaining amount %v", orderToRemove))
+		}
+
+		pendingOpenOrder := types.PendingOpenOrder{
+			RemainingQuantums: remainingAmount,
+			IsBuy:             orderToRemove.IsBuy(),
+			Subticks:          orderToRemove.GetOrderSubticks(),
+			ClobPairId:        orderToRemove.GetClobPairId(),
+		}
+
+		// Temporarily construct the subaccountOpenOrders with a single PendingOpenOrder.
+		subaccountOpenOrders := map[satypes.SubaccountId][]types.PendingOpenOrder{
+			orderIdToRemove.SubaccountId: {
+				pendingOpenOrder,
+			},
+		}
+
+		// TODO(DEC-1896): AddOrderToOrderbookCollatCheck should accept a single PendingOpenOrder as a
+		// parameter rather than the subaccountOpenOrders map.
+		_, successPerSubaccountUpdate := k.AddOrderToOrderbookCollatCheck(
```

Comment:
hmm I think this check is for the taker order. but maker orders can also fail collat check. 

if we are always doing the most pessimistic check here, maybe this is okay?

Reply:
Yeah I think this is a fine way to do it, cc @lucas-dydx to confirm